### PR TITLE
Major refactor

### PR DESCRIPTION
**Pull request by shermp**
I wasn't too happy with how... organic the code had gotten, so I decided to do a major refactor.

Quoting myself from the commit message:

- Switch USB audio code to C++.
- Add new HA and HAManager classes.
- Add PseudoAtomic library for proper variable syncing. This allows compiling in Release mode.
- The shared audio buffer is now a C++ class.
- Various other refactors.

No real improvements on the audio side. I am currently doing zero buffering, so the connection is unstable, but properly synced audio. Some ideas might be to detect underruns and insert silent packets to compensate, or figure out how to properly buffer without the stereo going out of sync.

I'd be interested to see your thoughts on this refactor @thewierdnut . I think it's better than what I had before, but I'm biased.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1,31 +1,37 @@
...
 {
+    // Init global shared variables
+    patom::PseudoAtomicInit();
+    audio_buff.init();
+    pcm_streaming = false;
+    encode_audio = false;
+
     // Get serial
     pico_get_unique_board_id_string(pico_uid, sizeof pico_uid);
 #ifdef ASHA_USB_SERIAL
```
Comment (thewierdnut) [inline src/main.cpp:37]: The `uid_suffix` is causing a compile error for me, because it doesn't have room for the null.

You can either do
```
const char uid_suffix[4] = {'-','C','D','C'};
```
or
```
const char uid_suffix[] = "-CDC";
memcpy(pico_uid + (sizeof(pico_uid) - sizeof(uid_suffix) - 2), uid_suffix, sizeof(uid_suffix)-1);
```

Comment (shermp) [reply src/main.cpp:37]: Ah, oops. I forgot that switching to C++ changes this.

What I was doing is legal in C.

Comment (shermp) [reply src/main.cpp:37]: FIxed. Thanks for pointing this out.

**Comment by thewierdnut** [comment]
Finally got my pico. What console settings are you using to connect to the compiled-in serial port? I've tried both minicom and kermit with various settings and haven't seen any output.

**Comment by shermp** [comment]
Settings are in the readme, but here they are:
- **Baud Rate:** 115200
- **Data Bits:** 8
- **Stop Bits:"** 1
- **Parity:** None
- **DTR** and **RTS** enabled

**Comment by shermp** [comment]
The other problem you might have is if you connect too late, you might not see any new output. That's why I prefer UART debugging myself, because my debug probe serial is always connected.

**Comment by shermp** [comment]
Oh, just checked, without hearing aids connected, there is no output. Maybe I need to add some scanning logs back in...

File: src/asha_bt.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+constexpr uint16_t buff_size_sdu = 161;
+
+/* Struct to hold data from a GAP advertisment report */
+struct AdvertisingReport {
+    bd_addr_t address;
+    uint8_t   type;
+    uint8_t   event_type;
+    uint8_t   address_type;
+    uint8_t   rssi;
```
Comment (thewierdnut) [inline src/asha_bt.hpp:33]: rssi is actually an int8_t. The values I have observed usually sit from -100 to about -30 or so.

File: src/asha_bt.hpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    bool is_hearing_aid;
```
Comment (thewierdnut) [inline src/asha_bt.hpp:35]: This is not being initialized, so its returning true for most advertisements.

Comment (thewierdnut) [reply src/asha_bt.hpp:35]: note that when I fix this, my devices are no longer reported as hearing aids anymore.

**Comment by thewierdnut** [comment]
I'm still having trouble during the service discovery part. I think you are trying to initiate the service discovery before the connection has been secured, so the hearing device ignores the request. 

I've attached an hci dump: [pico_output.log](https://github.com/user-attachments/files/16227754/pico_output.log)

File: src/asha_bt.hpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    ServiceDiscovery,
+    CharDiscovery,
+    ReadROP,
+    ReadPSM,
+    SubscribeASPNotification,
+    Finalizing,
+    Complete,
+};
+
+constexpr uint16_t buff_size_sdu = 161;
```
Comment (thewierdnut) [inline src/asha_bt.hpp:25]: This needs to be at least 167, to account for the l2cap header

**Comment by shermp** [comment]
Thanks for that log @thewierdnut 

It shows me that your device doesn't support LE Secure connections, so pairing fails.

> I'm still having trouble during the service discovery part. I think you are trying to initiate the service discovery before the connection has been secured, so the hearing device ignores the request.

Yeah, I think I know what I'm doing wrong there.

I've tried changing some stuff this morning, but it hasn't worked yet and I've got to go to work. I'll pick back up this evening. 

Sigh... I'm back to the pairing game...

**Comment by shermp** [comment]
In the meantime, could you try the following:

In `bt_main()`, could you change `sm_set_authentication_requirements(SM_AUTHREQ_SECURE_CONNECTION | SM_AUTHREQ_MITM_PROTECTION | SM_AUTHREQ_BONDING);` 
to 
`sm_set_authentication_requirements(SM_AUTHREQ_BONDING);

File: src/asha_bt.hpp (hunk 3 above)
Comment (shermp) [reply src/asha_bt.hpp]: Oops.

**Comment by shermp** [comment]
Just pushed those changes.

You can delete pairings again. Compile with `-DENABLE_PAIR_DEL=ON`.

Anyway, to work!

File: src/asha_bt.hpp (hunk 4 above)
Comment (shermp) [reply src/asha_bt.hpp:25]: As far as I'm aware, the SDU is the application data size, and does not contain l2cap headers. That's the PDU, and I believe btstack handles that automatically.

**Comment by shermp** [comment]
@thewierdnut I've made some changes to the security requirements. I'll be interested to see if this helps for you.

Your Starkey's are a PITA when it comes to pairing by the way :) They don't seem to support LE Secure Connections, wheras my Oticon's require it. So I've hacked in a "upgrade requirements on authentication failure and try again" mode.

**Comment by thewierdnut** [comment]
No luck:
```
[Pico-ASHA  INFO] MFI UUID discovered
    Flags: LE General Discoverable Mode; BR/EDR Not Supported; 
    Complete List of 128-bit Service Class UUIDs: 7D74F4BD-C74A-4431-862C-CCE884371592


[Pico-ASHA  SCAN] Got Ad Report from 9C:9C:1D:96:2F:5E
[Pico-ASHA  SCAN] RSSI: -62
[Pico-ASHA  INFO] Identity resolving succeeded
[Pico-ASHA  INFO] Connecting to address 9C:9C:1D:96:2F:5E
[Pico-ASHA  INFO] Device connected. Discovering ASHA service
[Pico-ASHA  INFO] Reencryption started
[Pico-ASHA  INFO] ASHA service not found. Continuing scanning
[Pico-ASHA ERROR] Reencryption failed with ERROR_CODE_PIN_OR_KEY_MISSING
[Pico-ASHA  INFO] Received disconnection event.
[Pico-ASHA  INFO] Expected disconnection
```

**Comment by thewierdnut** [comment]
Here is another hci dump with the latest code: [pico_output.log](https://github.com/user-attachments/files/16234645/pico_output.log)

I don't see the `Read by group type` request going out to discover the ASHA services.

**Comment by shermp** [comment]
It's still failing to properly pair - you can tell that by the reencryption failure.

Until you get it working, can you please enable `ENABLE_PAIR_DEL` to ensure a fresh pairing DB on every startup?

**Comment by shermp** [comment]
I think longer term, I'm going to need at least some level of control via USB serial, if not a GUI app to manage pairings etc.

**Comment by thewierdnut** [comment]
In that packet log, you can see in frame 226 where it requests Encryption, and in frame 258,  (201ms later) it responds, but before that in frame 240 is the client timeout stop message (5ms from the previous frame, 82ms from the encryption start, and 380ms from the connection start) , and it gives up listening. 

What timeout is this? The supervision timeout negotiated on connection start is 720 ms, which is much longer than this.

**Comment by shermp** [comment]
I think that timeout is a red herring. It appears to be an internal btstack GATT timeout, and is set to 30 seconds. I believe that message in the log is the timeout cancellation message.

I've changed the way pairing works so that it doesn't rely on the implicit pairing initiated by the GATT client API. Now, pairing happens before any attempt to discover services/characteristics.

**Comment by shermp** [comment]
Also, thank you for your patience as we attempt to get this working for your hearing aids. It is most appreciated.

**Comment by thewierdnut** [comment]
More progress. I still don't hear audio, but it appears to find the service and the characteristics: 
```
[Pico-ASHA  INFO] MFI UUID discovered
    Flags: LE General Discoverable Mode; BR/EDR Not Supported; 
    Complete List of 128-bit Service Class UUIDs: 7D74F4BD-C74A-4431-862C-CCE884371592


[Pico-ASHA  SCAN] Got Ad Report from 9C:9C:1D:96:2F:5E
[Pico-ASHA  SCAN] RSSI: -67
[Pico-ASHA  INFO] Hearing aid discovered with addr 9C:9C:1D:96:2F:5E. Connecting...
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Pairing started
[Pico-ASHA  INFO] Just Works requested
[Pico-ASHA  INFO] Pairing complete, success
[Pico-ASHA  INFO] Device paired. Discovering ASHA service
[Pico-ASHA  SCAN] Service Query: Service ID: 0x1800 : UUID: 00001800-0000-1000-8000-00805F9B34FB
[Pico-ASHA  SCAN] Service Query: Service ID: 0x1801 : UUID: 00001801-0000-1000-8000-00805F9B34FB
[Pico-ASHA  SCAN] Service Query: Service ID: 0x180a : UUID: 0000180A-0000-1000-8000-00805F9B34FB
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 896C9518-D4EA-11E1-AF45-58B035FEA743
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 5446A83A-D711-11E4-B9D6-1681E6B88EC1
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 5446E254-D711-11E4-B9D6-1681E6B88EC1
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: CED5FDB4-E645-0BB4-7A43-CABDF3197A77
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 544709DA-D711-11E4-B9D6-1681E6B88EC1
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 5446EC0E-D711-11E4-B9D6-1681E6B88EC1
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 48DDF118-EFD0-48FC-8F44-B9B8E17BE397
[Pico-ASHA  SCAN] Service Query: Service ID: 0xfdf0 : UUID: 0000FDF0-0000-1000-8000-00805F9B34FB
[Pico-ASHA  INFO] ASHA service found
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: D0B6DC42-03C8-457A-A347-ADE8D1C4E98D
[Pico-ASHA  SCAN] Service Query: Service ID: 0x180d : UUID: 0000180D-0000-1000-8000-00805F9B34FB
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 7D74F4BD-C74A-4431-862C-CCE884371592
[Pico-ASHA  SCAN] Service Query: Service ID: 0x0000 : UUID: 8341F2B4-C013-4F04-8197-C4CDB42E26DC
[Pico-ASHA  INFO] Got ROP Characteristic
[Pico-ASHA  INFO] Got ACP Characteristic
[Pico-ASHA  INFO] Got AUS Characteristic
[Pico-ASHA  INFO] Got VOL Characteristic
[Pico-ASHA  INFO] Got PSM Characteristic
[Pico-ASHA  INFO] ASHA characteristic discovery complete
[Pico-ASHA  INFO] Getting ReadOnlyProperties value
[Pico-ASHA  INFO] Completed value read of ReadOnlyProperties
[Pico-ASHA  INFO] Getting PSM value
[Pico-ASHA  INFO] PSM: 161
[Pico-ASHA  INFO] Completed value read of PSM
[Pico-ASHA  INFO] Read Only Properties
  Version:        1
  Side:           Right
  Mode:           Binaural
  CSIS:           No
  Manufacture ID: baxu
  LE CoC audio:   Yes
  Render delay:   160
  Supports 16KHz: Yes
  Supports 24KHz: No
[Pico-ASHA  INFO] AudioStatusPoint notification enabled.
```

File: src/hearing_aid.cpp [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+}
+
+void HA::ROP::print_values()
+{
+    LOG_INFO("Read Only Properties\n"
+        "  Version:        %d\n"
+        "  Side:           %s\n"
+        "  Mode:           %s\n"
+        "  CSIS:           %s\n"
+        "  Manufacture ID: %xxu\n"
```
Comment (thewierdnut) [inline src/hearing_aid.cpp:336]: Did you mean %02hx here?

File: src/asha_bt.cpp [hunk 6]
```diff
@@ -1240,58 +399,249 @@ static void sm_packet_handler (uint8_t packet_type, uint16_t channel, uint8_t *p
...
+    curr_scan.ha.state = HA::State::GATTConnected;
+    curr_scan.ha.l2cap_packet_handler = &l2cap_cbm_event_handler;
+    curr_scan.ha.gatt_packet_handler = &connected_gatt_event_handler;
+    auto& ha = ha_mgr.add(curr_scan.ha);
+    if (!ha) {
+        LOG_ERROR("Error adding this device.\n");
+        scan_state = ScanState::Scan;
+        return;
+    }
+    if (ha_mgr.set_complete()) {
```
Comment (thewierdnut) [inline src/asha_bt.cpp:664]: I wouldn't wait until both devices are available to start audio streaming. Just because devices report themselves as part of a pair doesn't mean that the user wants to use both of them. The stream tends to be much more stable with only one device, since it uses half the bandwidth and the devices aren't trying to sync. There are also use cases (such as somebody using a [cross device](https://github.com/thewierdnut/asha_pipewire_sink/issues/11), where stereo streaming is redundant because the hearing devices just mix them together into one ear anyways.

**Comment by thewierdnut** [comment]
Slightly more success when using both of my devices. It looks like you are seeing the same thing I do with asha_pipewire_sink, where if you connect to the right device first, then the left device, the left device tends to disconnect. I don't know why, and I don't observe this behavior with android. 
```
[Pico-ASHA  INFO] AudioStatusPoint notification enabled.
[Pico-ASHA  INFO] Connected to all aid(s) in set.
[Pico-ASHA  INFO] Left: Updating connection parameters
[Pico-ASHA  INFO] Right: Updating connection parameters
[Pico-ASHA  INFO] Right: Updated connection parameters
[Pico-ASHA  INFO] Right: Connecting to L2CAP
[Pico-ASHA  INFO] Left: Updated connection parameters
[Pico-ASHA  INFO] Left: Connecting to L2CAP
[Pico-ASHA ERROR] Left: L2CAP CoC failed with status code: 0x6a
[Pico-ASHA  INFO] Received disconnection event.
[Pico-ASHA ERROR] Disconnected with reason: 19
[Pico-ASHA  INFO] Left device disconnected.
```

**Comment by thewierdnut** [comment]
Latest hci trace: 
[pico_output.log](https://github.com/user-attachments/files/16249232/pico_output.log)

File: src/hearing_aid.cpp (hunk 5 above)
Comment (shermp) [reply src/hearing_aid.cpp:336]: Oh probably. I hate printf format specifiers :(

File: src/asha_bt.cpp (hunk 6 above)
Comment (shermp) [reply src/asha_bt.cpp:664]: Hmm yeah, good point.

File: CMakeLists.txt [hunk 7]
```diff
@@ -4,37 +4,47 @@ include(pico_sdk_import.cmake)
 
 project(pico_asha C CXX)
 set(CMAKE_C_STANDARD 11)
```
[the remark this answers is not in the document]
Comment (shermp) [reply CMakeLists.txt:6]: Sure, I'm not a cmake expert by any means, so if you have suggestions, suggest away!

**Comment by shermp** [comment]
> Slightly more success when using both of my devices. It looks like you are seeing the same thing I do with asha_pipewire_sink, where if you connect to the right device first, then the left device, the left device tends to disconnect. I don't know why, and I don't observe this behavior with android.

I've definitely had issues with unexpected disconnections, and I don't know why, because as you said, Android doesn't seem to have such troubles.

File: src/hearing_aid.cpp (hunk 5 above)
Comment (thewierdnut) [reply src/hearing_aid.cpp]: Sorry, that field is two bytes isn't it. you want `%04hx`

**Comment by thewierdnut** [comment]
So... spurious disconnect aside, when it tries to reconnect a second time it always gets this: 
```
[Pico-ASHA  INFO] MFI UUID discovered
    Flags: LE General Discoverable Mode; BR/EDR Not Supported; 
    Complete List of 128-bit Service Class UUIDs: 7D74F4BD-C74A-4431-862C-CCE884371592


[Pico-ASHA  SCAN] Got Ad Report from 9C:9C:1D:98:BE:82
[Pico-ASHA  SCAN] RSSI: -88
[Pico-ASHA  INFO] Identity resolving succeeded
[Pico-ASHA  INFO] Connecting to address 9C:9C:1D:98:BE:82
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Reencryption started
[Pico-ASHA ERROR] Reencryption failed with ERROR_CODE_PIN_OR_KEY_MISSING
```

**Comment by thewierdnut** [comment]
One thing that may help... You can go ahead and establish the CoC connection right after you have paired with the device. You don't have to wait until just before you want to start audio.

**Comment by thewierdnut** [comment]
Latest info logs. 
```
[Pico-ASHA  INFO] Connected to all aid(s) in set.
[Pico-ASHA  INFO] Right: Updating connection parameters
[Pico-ASHA  INFO] Left: Updating connection parameters
[Pico-ASHA  INFO] Left: Updated connection parameters
[Pico-ASHA  INFO] Left: Connecting to L2CAP
[Pico-ASHA  INFO] Right: Updated connection parameters
[Pico-ASHA  INFO] Right: Connecting to L2CAP
[Pico-ASHA ERROR] Right: L2CAP CoC failed with status code: 0x6a
[Pico-ASHA  INFO] Received disconnection event.
[Pico-ASHA ERROR] Disconnected with reason: 19
[Pico-ASHA  INFO] Right device disconnected.
[Pico-ASHA  INFO] MFI UUID discovered
[Pico-ASHA  INFO] Identity resolving succeeded
[Pico-ASHA  INFO] Connecting to address 9C:9C:1D:96:2F:5E
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Reencryption started
[Pico-ASHA ERROR] Reencryption failed with ERROR_CODE_PIN_OR_KEY_MISSING
[Pico-ASHA ERROR] Left: L2CAP CoC failed with status code: 0x6a
[Pico-ASHA  INFO] Received disconnection event.
[Pico-ASHA ERROR] Disconnected with reason: 19
[Pico-ASHA  INFO] Left device disconnected.
[Pico-ASHA  INFO] MFI UUID discovered
[Pico-ASHA  INFO] Identity resolving succeeded
[Pico-ASHA  INFO] Connecting to address 9C:9C:1D:98:BE:82
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Reencryption started
[Pico-ASHA ERROR] Reencryption failed with ERROR_CODE_PIN_OR_KEY_MISSING
```

**Comment by shermp** [comment]
This brings back nightmares of when I was attempting to get my HA's to properly pair...

That L2CAP error code seems to be `L2CAP_CONNECTION_BASEBAND_DISCONNECT` fwiw.

File: src/hearing_aid.cpp (hunk 5 above)
Comment (shermp) [reply src/hearing_aid.cpp]: Sigh... thank you.

**Comment by thewierdnut** [comment]
> That L2CAP error code seems to be `L2CAP_CONNECTION_BASEBAND_DISCONNECT` fwiw.

Yeah, the real error is the next one, 19 (0x13) which is ERROR_CODE_REMOTE_USER_TERMINATED_CONNECTION

The HAs didn't like something they saw, and just stopped responding.

**Comment by thewierdnut** [comment]
I also notice that the connection update doesn't seem to work. The meta event that returns after the update request still contains the old parameters.

**Comment by shermp** [comment]
Oh FFS! Your Starkeys keep on giving :)

Don't worry, I'm not upset. It's just perplexing.

**Comment by thewierdnut** [comment]
Latest log dump: 
[pico_output.log](https://github.com/user-attachments/files/16258115/pico_output.log)

**Comment by shermp** [comment]
I've made a few changes, I will be interested to see if that helps for your device @thewierdnut 

I'm just setting the default connection parameters up front, and doing away with the parameter update. 

I'm trying to perform the ASP notification subscription after the L2CAP connection is created (this is what Android appears to do, at least according to my android HCI snoop).

I also now attempt to delete a pairing on reencryption failure.

**Comment by thewierdnut** [comment]
I'll give it a try. 

A small note about the android setup state machine... As soon as the Gatt is connected, it triggers a call to UpdateBleConnParams, and if it ever gets a parameter update with the wrong parameters, it just calls UpdateBleConnParams again. It will actually connect to the gap socket as soon as the the encryption is complete and the characteristics are known (discovered or cached), which means the connection parameter update process and the service discovery can actually run in parallel. Once the connection parameters are correct and the socket is available, then it finally calls the ACP start.

**Comment by thewierdnut** [comment]
Still isn't happy 
[pico_output.log](https://github.com/user-attachments/files/16267951/pico_output.log)

**Comment by shermp** [comment]
Why oh why oh why? What am I doing wrong?

**Comment by thewierdnut** [comment]
I get the distinct idea that we are somehow overcomplicating this. I'm tempted to steal a btstack sample app and just statically connect to the psm via a mac address, just to see if it is stable. 

Also... what do you want to accomplish with this particular peer review? Do you want me to review code style? design? functionality? I think we have moved past that somewhat into troubleshooting a specific hearing aid.

**Comment by shermp** [comment]
> I get the distinct idea that we are somehow overcomplicating this. I'm tempted to steal a btstack sample app and just statically connect to the psm via a mac address, just to see if it is stable.

Yeah, maybe give that a go. Or feel free to hack away at my code.

> Also... what do you want to accomplish with this particular peer review? Do you want me to review code style? design? functionality? I think we have moved past that somewhat into troubleshooting a specific hearing aid.

Just whatever springs to mind really. This troubleshooting is fine, because it's a major refactor in draft state, so I have no worries making major changes. If you can suggest better ways of doing things, feel free, as this is my first ever microcontroller and bluetooth project, so I'm still learning stuff.

**Comment by shermp** [comment]
While I'm not caching the advertising packets, I decided to implement a cache of successfully connected HA's so that subsequent reconnections don't have to rediscover services and characteristics.

Anecdotally, the connection seems less stable after that change, but that could be anything, because bluetooth...

I think once we can get your Starkey's working, I will like to focus on stream stability. I think I will probably need to modify btstack a little to get access to the available outgoing credits.

**Comment by shermp** [comment]
Just did a quick bit of logging, it seems the audio streaming never falls behind the PCM stream, at least not to a significant degree. I tried logging with a difference of 8 and 4 packets, and the only time I saw any logging was when things had already gone wrong. I seem to have some state issues...

Maybe I partly just need to find some efficiencies as well. I keep having to remind myself this is running on a dual core ARM  Cortex M0+ with 256K of RAM. The M0+ is basically ARM's cheapest and slowest 32-bit CPU. 

I may also have to experiment with `__not_in_flash_func()` to force certain functions to always run from RAM, and not from flash using Execute In Place (XIP). (The XIP cache is 16K I believe).

**Comment by shermp** [comment]
Another option might be to investigate the SDK alarm API to schedule the streaming stuff instead of using an infinite loop.

**Comment by thewierdnut** [comment]
Here is the latest log output (with advertisements disabled) that you requested via slack:
```
[Pico-ASHA  INFO] BT ASHA star[Pico-ASHA  INFO] BT ASHA starting.
[Pico-ASHA  INFO] L2CAP Init.
[Pico-ASHA  INFO] SM Init.
[Pico-ASHA  INFO] GATT Client Init.
[Pico-ASHA  INFO] HCI power on.
[Pico-ASHA  INFO] Device DB is empty, allowing a 60s timeout
[Pico-ASHA  INFO] BTstack up and running on 28:CD:C1:0F:41:4A
[Pico-ASHA  INFO] Removing paired devices
[Pico-ASHA  INFO] Start scanning.
[Pico-ASHA  INFO] MFI UUID discovered
[Pico-ASHA  INFO] Hearing aid discovered with addr 9C:9C:1D:96:2F:5E. Connecting...
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Pairing started
[Pico-ASHA  INFO] Just Works requested
[Pico-ASHA  INFO] Pairing complete, success
[Pico-ASHA  INFO] Device paired. Discovering ASHA service
[Pico-ASHA  INFO] ASHA service found
[Pico-ASHA  INFO] Got ROP Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00b9  Value: 0x00ba  End: 0x00bb
[Pico-ASHA  INFO] Got ACP Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00bc  Value: 0x00bd  End: 0x00be
[Pico-ASHA  INFO] Got AUS Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00bf  Value: 0x00c0  End: 0x00c2
[Pico-ASHA  INFO] Got VOL Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00c3  Value: 0x00c4  End: 0x00c5
[Pico-ASHA  INFO] Got PSM Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00c6  Value: 0x00c7  End: 0x00c8
[Pico-ASHA  INFO] ASHA characteristic discovery complete
[Pico-ASHA  INFO] Getting ReadOnlyProperties value
[Pico-ASHA  INFO] Completed value read of ReadOnlyProperties
[Pico-ASHA  INFO] Getting PSM value
[Pico-ASHA  INFO] PSM: 161
[Pico-ASHA  INFO] Completed value read of PSM
[Pico-ASHA  INFO] Read Only Properties
  Version:        1
  Side:           Right
  Mode:           Binaural
  CSIS:           No
  Manufacture ID: 00ba
  LE CoC audio:   Yes
  Render delay:   160
  Supports 16KHz: Yes
  Supports 24KHz: No
[Pico-ASHA  INFO] MFI UUID discovered
[Pico-ASHA  INFO] Hearing aid discovered with addr 9C:9C:1D:98:BE:82. Connecting...
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Pairing started
[Pico-ASHA  INFO] Just Works requested
[Pico-ASHA  INFO] Pairing complete, success
[Pico-ASHA  INFO] Device paired. Discovering ASHA service
[Pico-ASHA  INFO] ASHA service found
[Pico-ASHA  INFO] Got ROP Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00b9  Value: 0x00ba  End: 0x00bb
[Pico-ASHA  INFO] Got ACP Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00bc  Value: 0x00bd  End: 0x00be
[Pico-ASHA  INFO] Got AUS Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00bf  Value: 0x00c0  End: 0x00c2
[Pico-ASHA  INFO] Got VOL Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00c3  Value: 0x00c4  End: 0x00c5
[Pico-ASHA  INFO] Got PSM Characteristic
[Pico-ASHA  INFO] Characteristic handles: Start: 0x00c6  Value: 0x00c7  End: 0x00c8
[Pico-ASHA  INFO] ASHA characteristic discovery complete
[Pico-ASHA  INFO] Getting ReadOnlyProperties value
[Pico-ASHA  INFO] Completed value read of ReadOnlyProperties
[Pico-ASHA  INFO] Getting PSM value
[Pico-ASHA  INFO] PSM: 161
[Pico-ASHA  INFO] Completed value read of PSM
[Pico-ASHA  INFO] Read Only Properties
  Version:        1
  Side:           Left
  Mode:           Binaural
  CSIS:           No
  Manufacture ID: 00ba
  LE CoC audio:   Yes
  Render delay:   160
  Supports 16KHz: Yes
  Supports 24KHz: No
[Pico-ASHA  INFO] Connected to all aid(s) in set.
[Pico-ASHA  INFO] Right: Connecting to L2CAP
[Pico-ASHA  INFO] Left: Connecting to L2CAP
[Pico-ASHA ERROR] Right: L2CAP CoC failed with status code: 0x6a
[Pico-ASHA  INFO] Received disconnection event.
[Pico-ASHA ERROR] Disconnected with reason: 19
[Pico-ASHA  INFO] Right device disconnected.
[Pico-ASHA  INFO] MFI UUID discovered
[Pico-ASHA  INFO] Identity resolving succeeded
[Pico-ASHA  INFO] Connecting to address 9C:9C:1D:96:2F:5E
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Reencryption started
[Pico-ASHA ERROR] Reencryption failed with ERROR_CODE_PIN_OR_KEY_MISSING
[Pico-ASHA  INFO] Removing paired device with address 9C:9C:1D:96:2F:5E
[Pico-ASHA  INFO] Found. Removing
[Pico-ASHA ERROR] Left: L2CAP CoC failed with status code: 0x6a
[Pico-ASHA  INFO] Received disconnection event.
[Pico-ASHA ERROR] Disconnected with reason: 19
[Pico-ASHA  INFO] Left device disconnected.
[Pico-ASHA  INFO] MFI UUID discovered
[Pico-ASHA  INFO] Identity resolving succeeded
[Pico-ASHA  INFO] Connecting to address 9C:9C:1D:98:BE:82
[Pico-ASHA  INFO] Device connected. Attempt pairing
[Pico-ASHA  INFO] Reencryption started
[Pico-ASHA ERROR] Reencryption failed with ERROR_CODE_PIN_OR_KEY_MISSING
[Pico-ASHA  INFO] Removing paired device with address 9C:9C:1D:98:BE:82
[Pico-ASHA  INFO] Found. Removing
```

**Comment by shermp** [comment]
Thanks, so it seems it's fine up until attempting to create the l2cap channel, at which point it has a melt-down, and for the cherry on top, it refuses to pair after reconnecting.

The l2cap error is `L2CAP_CONNECTION_BASEBAND_DISCONNECT`. Looking at the btstack source I see the following:
```c
#ifdef ENABLE_L2CAP_LE_CREDIT_BASED_FLOW_CONTROL_MODE
static void l2cap_handle_hci_le_disconnect_event(l2cap_channel_t * channel){
    if (l2cap_send_open_failed_on_hci_disconnect(channel)){
        l2cap_cbm_emit_channel_opened(channel, L2CAP_CONNECTION_BASEBAND_DISCONNECT);
    } else {
        l2cap_emit_channel_closed(channel);
    }
    l2cap_free_channel_entry(channel);
}
#endif
```

The error is based on what `l2cap_send_open_failed_on_hci_disconnect(channel)` returns, so lets look at that:

```c
static int l2cap_send_open_failed_on_hci_disconnect(l2cap_channel_t * channel){
    // open cannot fail for for incoming connections
    if (channel->state_var & L2CAP_CHANNEL_STATE_VAR_INCOMING) return 0;

    // check state
    switch (channel->state){
        case L2CAP_STATE_WILL_SEND_CREATE_CONNECTION:
        case L2CAP_STATE_WAIT_CONNECTION_COMPLETE:
        case L2CAP_STATE_WAIT_REMOTE_SUPPORTED_FEATURES:
        case L2CAP_STATE_WAIT_OUTGOING_SECURITY_LEVEL_UPDATE:
        case L2CAP_STATE_WAIT_CLIENT_ACCEPT_OR_REJECT:
        case L2CAP_STATE_WAIT_OUTGOING_EXTENDED_FEATURES:
        case L2CAP_STATE_WAIT_CONNECT_RSP:
        case L2CAP_STATE_CONFIG:
        case L2CAP_STATE_WILL_SEND_CONNECTION_REQUEST:
        case L2CAP_STATE_WILL_SEND_LE_CONNECTION_REQUEST:
        case L2CAP_STATE_WAIT_LE_CONNECTION_RESPONSE:
        case L2CAP_STATE_EMIT_OPEN_FAILED_AND_DISCARD:
            return 1;

        case L2CAP_STATE_OPEN:
        case L2CAP_STATE_CLOSED:
        case L2CAP_STATE_WAIT_INCOMING_EXTENDED_FEATURES:
        case L2CAP_STATE_WAIT_DISCONNECT:
        case L2CAP_STATE_WILL_SEND_CONNECTION_RESPONSE_INSUFFICIENT_SECURITY:
        case L2CAP_STATE_WILL_SEND_CONNECTION_RESPONSE_DECLINE:
        case L2CAP_STATE_WILL_SEND_CONNECTION_RESPONSE_ACCEPT:
        case L2CAP_STATE_WILL_SEND_DISCONNECT_REQUEST:
        case L2CAP_STATE_WILL_SEND_DISCONNECT_RESPONSE:
        case L2CAP_STATE_WILL_SEND_LE_CONNECTION_RESPONSE_DECLINE:
        case L2CAP_STATE_WILL_SEND_LE_CONNECTION_RESPONSE_ACCEPT:
        case L2CAP_STATE_INVALID:
        case L2CAP_STATE_WAIT_INCOMING_SECURITY_LEVEL_UPDATE:
            return 0;

        default:
            // get a "warning" about new states
            btstack_assert(false);
            return 0;
    }
}
```

In other words, it could be any of the first 12 l2cap states. Very helpful btstack...

**Comment by thewierdnut** [comment]
If you look at the raw hci traffic, you see this just prior to the logged disconnect:
image

**Comment by shermp** [comment]
@thewierdnut I've added a basic timestamp to logs, that should give more context, without majorly screwing up timings like a HCI dump does.

**Comment by shermp** [comment]
Apologies for the lack of updates. I started trying some stuff, and then my plans were all derailed early this week.

**Comment by thewierdnut** [comment]
No worries. Life happens. I haven't had time to work on it either.

**Comment by shermp** [comment]
New Pico SDK is out. I think I'm just gonna merge this as-is, then start hopefully making stability improvements.
